rust/lance-core/src/utils/backoff.rs

```diff
@@ -73,6 +77,64 @@ impl Backoff {
     }
 }
 
+/// SlotBackoff is a backoff strategy that uses a slot-based approach.
+///
+/// This is for when the cause of the failure is concurrency itself and that
+/// the attempts take about the same amount of time.
+///
+/// Say you have N attempts to do something. We don't know there are N ahead of
+/// time. We start guessing with 4 slots:
+///
+/// | 1, 2, 3 | 4, 5, 6 | 7, 8, 9 | 10 |
+///
+/// Each slot can have one success, so we can eliminate 3, 6, 9, and 10. In the
+/// next round, we will use twice as many slots (8):
+///
+/// | 1 | 2 | 4 | 5 | 7 | 8 | ... |
+///
+/// Optimally, that should be 16 total attempts. For a 1s unit, the retry times
+/// are:
+/// * Round 1: 0 - 3s
+/// * Round 2: 0 - 7s
+/// * Round 3: 0 - 15s
+/// * ...
+pub struct SlotBackoff {
+    base: u32,
+    unit: u32,
+    starting_i: u32,
+    attempt: u32,
+    rng: rand::rngs::SmallRng,
+}
+
+impl Default for SlotBackoff {
+    fn default() -> Self {
+        Self {
+            base: 2,
+            unit: 50,
+            starting_i: 2, // start with 4 slots
+            attempt: 0,
+            rng: rand::rngs::SmallRng::from_entropy(),
+        }
+    }
+}
+
+impl SlotBackoff {
+    pub fn with_unit(self, unit: u32) -> Self {
+        Self { unit, ..self }
+    }
+
+    pub fn attempt(&self) -> u32 {
+        self.attempt
+    }
+
+    pub fn next_backoff(&mut self) -> Duration {
+        let num_slots = self.base.saturating_pow(self.attempt + self.starting_i);
+        let slot_i = self.rng.gen_range(0..num_slots);
+        self.attempt += 1;
+        Duration::from_millis((slot_i * self.unit) as u64)
+    }
```

Comment:
This is pretty cool.  Did you come up with this or is it based on something?

Reply:
I came up with this, when I realized that exponential backoff wouldn't really help here. I haven't found anything similar when googling, but could just be missing the right keywords.